### PR TITLE
Add new module for elasticache parameter group.

### PR DESCRIPTION
**Pull request by akurniawan**
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
elasticache_parameter_group module

##### ANSIBLE VERSION
```
2.2.0
```

##### SUMMARY
This is new module to create parameter group for elasticache in AWS

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    existing_pg = get_current_pg(es_client, module)
+
+    if state == "absent":
+        response, changed = remove_elasticache_pg(
+            es_client, existing_pg, module)
+    elif state == "present":
+        response, changed = upsert_elasticache_pg(
+            es_client, existing_pg, module)
+
+    if changed:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py]: you can just `module.exit_json(changed=changed, parameter_groups=response)`

File: lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    HAS_BOTO = False
+
+"""
+Helper functions
+"""
+
+
+def get_es_client(params):
+    """Get elasticache client"""
+    session = boto3.Session(aws_access_key_id=params["aws_access_key"],
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py]: we should be using boto3_conn for all connections. For example, security_token is missing here. boto3_conn avoids you having to worry about that.

File: lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+- elasticache_parameter_group:
+    state: "absent"
+    pg_name: "testing-pg"
+"""
+
+try:
+    import boto3
+    import botocore
+    HAS_BOTO = True
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py]: should be HAS_BOTO3

**Comment by zimbatm** [comment]
Please remove me from the bot mentions. The linked help page doesn't give me a way to do it myself.

**Comment by akurniawan** [comment]
Hi @zimbatm, how exactly can I do that?

File: lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+DOCUMENTATION = """
+---
+module: elasticache_parameter_group
+short_description: Manage elasticache parameter group
+description:
+  - Manage elasticache parameter group
+  - Has the capabilities to create/destroy parameter group
+version_added: "2.2"
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py]: Should be 2.3

File: lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        before_name = bp["ParameterName"]
+        before_value = bp["ParameterValue"]
+        if before_name in after_params:
+            if str(before_value) != str(after_params[before_name]):
+                return True
+    return False
+
+
+def convert_params_format(params):
+    result = []
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py]: This method can be simplified to:

```
return [{"ParameterName": str(k), "ParameterValue": str(v)} for (k,v) in params.items]
```

File: lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    if state == "absent":
+        response, changed = remove_elasticache_pg(
+            es_client, existing_pg, module)
+    elif state == "present":
+        response, changed = upsert_elasticache_pg(
+            es_client, existing_pg, module)
+
+    module.exit_json(changed=changed, parameter_groups=response)
+
+from ansible.module_utils.basic import *
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py]: Should be changed to new style module imports nearer the top of the file:

```
from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.ec2 import ec2_argument_spec, get_aws_connection_info, boto3_conn
```

**Review by willthames — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py (hunk 6 above)
Comment (willthames) [reply lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py]: This is still outstanding

File: lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py (hunk 5 above)
Comment (willthames) [reply lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py]: This is a suggestion, not a blocker.

File: lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py (hunk 6 above)
Comment (akurniawan) [reply lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py]: should I put the imports to the very beginning of the file or above DOCUMENTATION variable?

Comment (willthames) [reply lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py]: I put mine after the comments just above the documentation variable as then it passes pep8.

File: lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+try:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py]: flake8 will moan about this - and apologies for not being clearer. Leave this section where it was, and then everything should be fine. One thing to note while I'm in the area,but it's not a blocker, is that flake8 will complain less if you add HAS_BOTO3 to the list of imports and don't import boto3 (in the `except ImportError` part, just `pass`, as you'll handle that  using HAS_BOTO3 later)

Comment (akurniawan) [reply lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py]: I don't quite follow this part 
> add HAS_BOTO3 to the list of imports and don't import boto3. 

Is this what you mean?
try:
    import botocore
    HAS_BOTO3 = True
except ImportError:
    pass

How can I tell if boto3 is really installed?

Comment (willthames) [reply lib/ansible/modules/cloud/amazon/elasticache_parameter_group.py]:
```
from ansible.module_utils.ec2 import HAS_BOTO3

try:
    import botocore
except ImportError:
    pass  # will be caught by imported HAS_BOTO3
```

**Comment by willthames** [comment]
I'm sorry, I've given bad advice - ansible/proposals#50 now means that, in contradiction to what standard flake8 says, the imports should be after all the documentation. This has been a little bit of a moving target during my reviews of this PR, but I still apologise for the misinformation. Should be an easy fix though.

**Comment by s-hertel** [comment]
Closing because #21023 was merged. If you think there is missing functionality, please let me know.

**Pull request closed without merging**